NDBC Metocean data

akeeste opened the pull request:
This PR adds the ability to read continuous wind data to the NDBC I/O module. It does the following:
- [x] adds the "cwind" variable to each NDBC function documentation
- [x] jupyter notebook example (`wind_example`) to read and visualize wind data
- [x] adds test to read continuous wind data files, both with and without units
- [x] adds a check in ``waves.io.ndbc.request_data()`` to catch and remove units from numerical data.

ssolson commented on examples/wind_example.ipynb:
```diff
@@ -0,0 +1,925 @@
+    "    end_time = '2018-01-11 ' + str(hour) + ':30:00'\n",
+    "    mask = (daily_data.index >= start_time) * (daily_data.index < end_time)\n",
+    "    hourly_data = daily_data[mask]\n",
+    "    hourly_mean_speeds.append(np.mean(hourly_data.WSPD))"
```
I believe this whole code block is simply :
`daily_data.resample('h').mean()`

akeeste replied:
Thanks! I wasn't aware of this dataFrame method yet. Updated this to use ``resample()``

ssolson commented on examples/wind_example.ipynb:
```diff
@@ -0,0 +1,925 @@
+    "ax.set_xlabel('Time (h)')\n",
+    "ax.set_ylabel('Speed (m/s)')\n",
+    "ax.set_title('Hourly mean wind speeds on January 11, 2018')\n",
+    "line = ax.plot(hours,hourly_mean_speeds,'o')"
```
what do you think about adding gridlines to the plot?

akeeste replied:
Good call, it makes the plot easier to read

ssolson commented on examples/wind_example.ipynb:
```diff
@@ -0,0 +1,925 @@
+   ],
+   "source": [
+    "monthly_data = ndbc_data['2018']['2018-01']\n",
+    "daily_data = monthly_data['2018-01-10']\n",
```
Because you need to repeat the month ('01') in the monthly data and daily data I would suggest making this yearly data then daily data or simply:
`daily_data = ndbc_data['2018']['2018-01-10']`

akeeste replied:
Along with renaming these variables for clarity, I made Jan_data and Jan_11_data reference the yearly dataFrame directly

ssolson commented on examples/wind_example.ipynb:
```diff
@@ -0,0 +1,925 @@
+    "daily_data = monthly_data['2018-01-10']\n",
+    "\n",
+    "# Remove bad speed data\n",
+    "mean_speed_data = daily_data[daily_data.WDIR < 999]\n",
```
This operation is okay. I think this would be better:
`daily_data['WDIR'] = daily_data['WDIR'].replace({999.0:np.NaN})`

I think you should mention the different NaN values when they are first shown up above on the initial import. Maybe replace them at that time.

Also, you change the name to a mean but do not take a mean, is this correct?

akeeste replied:
Good call, I'll add a section about replacing the NaN values with np.NaN and do it up front before creating separate daily and monthly datasets.

ssolson commented on examples/wind_example.ipynb:
```diff
@@ -0,0 +1,925 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjAAAAHnCAYAAACmImdlAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAEAAElEQVR4nOydd3xT5f7HP092s7uyuvcedEBLmQqKoBcRUMQreAFBxPUT9brnVa974d6owHWggIoTkFVaWqB0QindbbrT3WY9vz9KK6O7aZO05/16nRc0OefJN2l6zud8J6GUgoGBgYGBgYHBnmBZ2wAGBgYGBgYGhqHCCBgGBgYGBgYGu4MRMAwMDAwMDAx2ByNgGBgYGBgYGOwORsA ...
```
Again I would recommend using the replace function here

ssolson commented on examples/wind_example.ipynb:
```diff
@@ -0,0 +1,925 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjAAAAHnCAYAAACmImdlAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAEAAElEQVR4nOydd3xT5f7HP092s7uyuvcedEBLmQqKoBcRUMQreAFBxPUT9brnVa974d6owHWggIoTkFVaWqB0QindbbrT3WY9vz9KK6O7aZO05/16nRc0OefJN2l6zud8J6GUgoGBgYGBgYHBnmBZ2wAGBgYGBgYGhqHCCBgGBgYGBgYGu4MRMAwMDAwMDAx2ByNgGBgYGBgYGOwORsA ...
```
I don't really think `daily_data` is a good name here. Shouldn't it be `Jan_data`?

akeeste replied:
Agreed, I renamed the daily/monthly variables to ``Jan_data`` and ``Jan_11_data`` for clarity

ssolson reviewed (commented):
Hey Adam great job on this. I have just a couple minor comments to kick off the review.

Overall I think you explain the use of the NDBC model much better than I did in the Pacwave example. Therefore I think I will remove my description of using wave.io.ndbc and point to this example.

@rpauly18 this example is the basis of the metocean module. It is currently called `wind_example`. I think we could leave this as wind example for now or rename to `metocean`. If we don't like either of those for now this could be renamed NDBC example as we do not have one specifically for that wave.io submodule. What is your thoughts/ preferences?

rpauly18 commented:
@ssolson I prefer "metocean" over "wind"

akeeste commented:
@ssolson @rpauly18 
I made the requested changes to this PR including:
- [x] Move graphics import from the example to the ndbc module
- [x] Fix capitalization and other minor grammar
- [x] Add grid lines to time series plot
- [x] Renamed the variables to Jan_data and Jan_11_data for clarity, and replaced the hour sampling with a call to resample() for simplicity
- [x] Replaced 99, 999, 9999 with NaN up front and clarified the need to do so

akeeste commented on examples/wind_example.ipynb:
```diff
@@ -0,0 +1,925 @@
+   "source": [
+    "from mhkit.wave.io import ndbc\n",
+    "from mhkit.tidal.graphics import plot_rose \n",
+    "from mhkit.tidal.graphics import plot_joint_probability_distribution\n",
```
I moved these imports out of the example and into the ndbc module, this seemed most similar to how the tidal module inherited from the river module
